Disk with file_id and size causes plugin panic

**Describe the bug**
I have the problem with a ubuntu cloud image (referenced via file_id) and the size for the corresponding disk.
When no size is defined, the space of the disk is as much as in the image, then no additional software installation is possible.
When i add the size element to the disk, the vm will be created completely and is succesfully configured (with the needed size), but the plugin crashes.

Following Ressource configuration causes the plugin panic
```
+ resource "proxmox_virtual_environment_vm" "virtual-machine" {
      + acpi                    = true
      + bios                    = "ovmf"
      + boot_order              = []
      + id                      = (known after apply)
      + ipv4_addresses          = (known after apply)
      + ipv6_addresses          = (known after apply)
      + keyboard_layout         = "de"
      + mac_addresses           = (known after apply)
      + machine                 = "pc"
      + migrate                 = true
      + name                    = "DE-FRA2-JOEK1"
      + network_interface_names = (known after apply)
      + node_name               = "mgmthost1"
      + on_boot                 = true
      + reboot                  = false
      + scsi_hardware           = "virtio-scsi-pci"
      + started                 = true
      + tablet_device           = true
      + template                = false
      + timeout_clone           = 1800
      + timeout_create          = 1800
      + timeout_migrate         = 1800
      + timeout_move_disk       = 1800
      + timeout_reboot          = 120
      + timeout_shutdown_vm     = 120
      + timeout_start_vm        = 1800
      + timeout_stop_vm         = 300
      + vm_id                   = (known after apply)

      + cpu {
          + architecture = "x86_64"
          + cores        = 2
          + hotplugged   = 0
          + numa         = false
          + sockets      = 1
          + type         = "qemu64"
          + units        = 1024
        }

      + disk {
          + cache             = "none"
          + datastore_id      = "local-zfs"
          + discard           = "ignore"
          + file_format       = (known after apply)
          + file_id           = "local:iso/jammy-server-cloudimg-amd64-disk-kvm.img"
          + interface         = "virtio0"
          + iothread          = false
          + path_in_datastore = (known after apply)
          + size              = 5
          + ssd               = false
        }

      + efi_disk {
          + datastore_id      = "local-zfs"
          + file_format       = "raw"
          + pre_enrolled_keys = true
          + type              = "4m"
        }

      + initialization {
          + datastore_id         = "local"
          + interface            = "scsi2"
          + network_data_file_id = "local:snippets/cloud-init-network.yml"
          + user_data_file_id    = "local:snippets/cloud-init-user.yml"
        }

      + memory {
          + dedicated = 4096
          + floating  = 0
          + shared    = 0
        }

      + network_device {
          + bridge      = "vmbr0"
          + enabled     = true
          + firewall    = false
          + mac_address = (known after apply)
          + model       = "virtio"
          + mtu         = 0
          + queues      = 0
          + rate_limit  = 0
          + vlan_id     = 0
        }

      + operating_system {
          + type = "l26"
        }

      + serial_device {
          + device = "socket"
        }

      + vga {
          + enabled = true
          + memory  = 16
          + type    = "serial0"
        }
    }

```

Panic Stack Trace:
```
proxmox_virtual_environment_vm.virtual-machine["DE-FRA2-JOEK1"]: Still creating... [10s elapsed]
╷
│ Error: Plugin did not respond
│ 
│ The plugin encountered an error, and failed to respond to the
│ plugin6.(*GRPCProvider).ApplyResourceChange call. The plugin logs may
│ contain more details.
╵

Stack trace from the terraform-provider-proxmox_v0.40.0 plugin:

panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xcabf20]

goroutine 117 [running]:
github.com/bpg/terraform-provider-proxmox/proxmoxtf/resource.vmReadCustom({0x103ed48, 0xc000718af0}, 0xc00059fd00, {0xe14960?, 0xc0005d2000?}, 0x0?, 0xc000134000, 0x424a3c?)
	github.com/bpg/terraform-provider-proxmox/proxmoxtf/resource/vm.go:4137 +0xd160
github.com/bpg/terraform-provider-proxmox/proxmoxtf/resource.vmRead({0x103ed48, 0xc000718af0}, 0xc00059fd00, {0xe14960?, 0xc0005d2000?})
	github.com/bpg/terraform-provider-proxmox/proxmoxtf/resource/vm.go:3789 +0x791
github.com/bpg/terraform-provider-proxmox/proxmoxtf/resource.vmCreateStart({0x103ed48, 0xc000718af0}, 0xc00059fd00, {0xe14960?, 0xc0005d2000?})
	github.com/bpg/terraform-provider-proxmox/proxmoxtf/resource/vm.go:2997 +0x625
github.com/bpg/terraform-provider-proxmox/proxmoxtf/resource.vmCreateCustomDisks({0x103ed48, 0xc000718af0}, 0xc00059fd00, {0xe14960?, 0xc0005d2000?})
	github.com/bpg/terraform-provider-proxmox/proxmoxtf/resource/vm.go:2951 +0x1876
github.com/bpg/terraform-provider-proxmox/proxmoxtf/resource.vmCreateCustom({0x103ed48, 0xc000718af0}, 0x1644c40?, {0xe14960?, 0xc0005d2000?})
	github.com/bpg/terraform-provider-proxmox/proxmoxtf/resource/vm.go:2800 +0x3145
github.com/bpg/terraform-provider-proxmox/proxmoxtf/resource.vmCreate({0x103ed48, 0xc000718af0}, 0x0?, {0xe14960, 0xc0005d2000})
	github.com/bpg/terraform-provider-proxmox/proxmoxtf/resource/vm.go:1649 +0x94
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).create(0xc0002fd260, {0x103eca0, 0xc000157410}, 0xd?, {0xe14960, 0xc0005d2000})
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.30.0/helper/schema/resource.go:778 +0x11b
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).Apply(0xc0002fd260, {0x103eca0, 0xc000157410}, 0xc0004a96c0, 0xc00059fb80, {0xe14960, 0xc0005d2000})
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.30.0/helper/schema/resource.go:909 +0xa7e
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ApplyResourceChange(0xc0000c3440, {0x103eca0?, 0xc000157260?}, 0xc000594eb0)
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.30.0/helper/schema/grpc_provider.go:1072 +0xdbc
github.com/hashicorp/terraform-plugin-mux/tf5to6server.v5tov6Server.ApplyResourceChange({{0x1046280?, 0xc0000c3440?}}, {0x103eca0, 0xc000157260}, 0x0?)
	github.com/hashicorp/terraform-plugin-mux@v0.12.0/tf5to6server/tf5to6server.go:37 +0x54
github.com/hashicorp/terraform-plugin-mux/tf6muxserver.(*muxServer).ApplyResourceChange(0x103ecd8?, {0x103eca0?, 0xc000156f60?}, 0xc000594e60)
	github.com/hashicorp/terraform-plugin-mux@v0.12.0/tf6muxserver/mux_server_ApplyResourceChange.go:36 +0x193
github.com/hashicorp/terraform-plugin-go/tfprotov6/tf6server.(*server).ApplyResourceChange(0xc0001d3900, {0x103eca0?, 0xc000156300?}, 0xc000718000)
	github.com/hashicorp/terraform-plugin-go@v0.19.1/tfprotov6/tf6server/server.go:857 +0x56a
github.com/hashicorp/terraform-plugin-go/tfprotov6/internal/tfplugin6._Provider_ApplyResourceChange_Handler({0xe8bde0?, 0xc0001d3900}, {0x103eca0, 0xc000156300}, 0xc00059ec00, 0x0)
	github.com/hashicorp/terraform-plugin-go@v0.19.1/tfprotov6/internal/tfplugin6/tfplugin6_grpc.pb.go:467 +0x169
google.golang.org/grpc.(*Server).processUnaryRPC(0xc0001ce1e0, {0x103eca0, 0xc000156240}, {0x1044830, 0xc0003b6820}, 0xc000237680, 0xc0003ace40, 0x1637a78, 0x0)
	google.golang.org/grpc@v1.59.0/server.go:1343 +0xe03
google.golang.org/grpc.(*Server).handleStream(0xc0001ce1e0, {0x1044830, 0xc0003b6820}, 0xc000237680)
	google.golang.org/grpc@v1.59.0/server.go:1737 +0xc4c
google.golang.org/grpc.(*Server).serveStreams.func1.1()
	google.golang.org/grpc@v1.59.0/server.go:986 +0x86
created by google.golang.org/grpc.(*Server).serveStreams.func1 in goroutine 25
	google.golang.org/grpc@v1.59.0/server.go:997 +0x145

Error: The terraform-provider-proxmox_v0.40.0 plugin crashed!
```

**Expected behavior**
no crash, normal exit of the plugin

**Additional context**
Have tested with pc and q35 machine type.

- clustered Proxmox:
- Provider version (ideally it should be the latest version): 0.40.0
- Terraform version: v1.4.6
- OS (where you run Terraform from): alpine linux in docker container
- Debug logs (`TF_LOG=DEBUG terraform apply`): attached
[terraform proxmox provider.log](https://github.com/bpg/terraform-provider-proxmox/files/13657217/terraform.proxmox.provider.log)

--- a/proxmox/nodes/vms/vms_types.go
+++ b/proxmox/nodes/vms/vms_types.go
@@ -187,8 +187,8 @@ type CustomStorageDevice struct {
 }
 
 // PathInDatastore returns path part of FileVolume or nil if it is not yet allocated.
-func (r CustomStorageDevice) PathInDatastore() *string {
-	probablyDatastoreID, pathInDatastore, hasDatastoreID := strings.Cut(r.FileVolume, ":")
+func (d CustomStorageDevice) PathInDatastore() *string {
+	probablyDatastoreID, pathInDatastore, hasDatastoreID := strings.Cut(d.FileVolume, ":")
 	if !hasDatastoreID {
 		// when no ':' separator is found, 'Cut' places the whole string to 'probablyDatastoreID',
 		// we want it in 'pathInDatastore' (as it is absolute filesystem path)
@@ -215,8 +215,8 @@ func (r CustomStorageDevice) PathInDatastore() *string {
 }
 
 // IsOwnedBy returns true, if CustomStorageDevice is owned by given VM. Not yet allocated volumes are not owned by any VM.
-func (r CustomStorageDevice) IsOwnedBy(vmID int) bool {
-	pathInDatastore := r.PathInDatastore()
+func (d CustomStorageDevice) IsOwnedBy(vmID int) bool {
+	pathInDatastore := d.PathInDatastore()
 	if pathInDatastore == nil {
 		// not yet allocated volume, consider disk not owned by any VM
 		// NOTE: if needed, create IsOwnedByOtherThan(vmId) instead of changing this return value.
@@ -234,6 +234,12 @@ func (r CustomStorageDevice) IsOwnedBy(vmID int) bool {
 	}
 
 	return false
+}
+
+// IsCloudInitDrive returns true, if CustomStorageDevice is a cloud-init drive.
+func (d CustomStorageDevice) IsCloudInitDrive(vmID int) bool {
+	return d.Media != nil && *d.Media == "cdrom" &&
+		strings.Contains(d.FileVolume, fmt.Sprintf("vm-%d-cloudinit", vmID))
 }
 
 // CustomStorageDevices handles QEMU SATA device parameters.

--- a/proxmox/nodes/vms/vms_types.go
+++ b/proxmox/nodes/vms/vms_types.go
@@ -1160,73 +1160,73 @@ func (r CustomStartupOrder) EncodeValues(key string, v *url.Values) error {
 }
 
 // EncodeValues converts a CustomStorageDevice struct to a URL vlaue.
-func (r CustomStorageDevice) EncodeValues(key string, v *url.Values) error {
+func (d CustomStorageDevice) EncodeValues(key string, v *url.Values) error {
 	values := []string{
-		fmt.Sprintf("file=%s", r.FileVolume),
+		fmt.Sprintf("file=%s", d.FileVolume),
 	}
 
-	if r.AIO != nil {
-		values = append(values, fmt.Sprintf("aio=%s", *r.AIO))
+	if d.AIO != nil {
+		values = append(values, fmt.Sprintf("aio=%s", *d.AIO))
 	}
 
-	if r.BackupEnabled != nil {
-		if *r.BackupEnabled {
+	if d.BackupEnabled != nil {
+		if *d.BackupEnabled {
 			values = append(values, "backup=1")
 		} else {
 			values = append(values, "backup=0")
 		}
 	}
 
-	if r.BurstableReadSpeedMbps != nil {
-		values = append(values, fmt.Sprintf("mbps_rd_max=%d", *r.BurstableReadSpeedMbps))
+	if d.BurstableReadSpeedMbps != nil {
+		values = append(values, fmt.Sprintf("mbps_rd_max=%d", *d.BurstableReadSpeedMbps))
 	}
 
-	if r.BurstableWriteSpeedMbps != nil {
-		values = append(values, fmt.Sprintf("mbps_wr_max=%d", *r.BurstableWriteSpeedMbps))
+	if d.BurstableWriteSpeedMbps != nil {
+		values = append(values, fmt.Sprintf("mbps_wr_max=%d", *d.BurstableWriteSpeedMbps))
 	}
 
-	if r.Format != nil {
-		values = append(values, fmt.Sprintf("format=%s", *r.Format))
+	if d.Format != nil {
+		values = append(values, fmt.Sprintf("format=%s", *d.Format))
 	}
 
-	if r.MaxReadSpeedMbps != nil {
-		values = append(values, fmt.Sprintf("mbps_rd=%d", *r.MaxReadSpeedMbps))
+	if d.MaxReadSpeedMbps != nil {
+		values = append(values, fmt.Sprintf("mbps_rd=%d", *d.MaxReadSpeedMbps))
 	}
 
-	if r.MaxWriteSpeedMbps != nil {
-		values = append(values, fmt.Sprintf("mbps_wr=%d", *r.MaxWriteSpeedMbps))
+	if d.MaxWriteSpeedMbps != nil {
+		values = append(values, fmt.Sprintf("mbps_wr=%d", *d.MaxWriteSpeedMbps))
 	}
 
-	if r.Media != nil {
-		values = append(values, fmt.Sprintf("media=%s", *r.Media))
+	if d.Media != nil {
+		values = append(values, fmt.Sprintf("media=%s", *d.Media))
 	}
 
-	if r.Size != nil {
-		values = append(values, fmt.Sprintf("size=%s", *r.Size))
+	if d.Size != nil {
+		values = append(values, fmt.Sprintf("size=%s", *d.Size))
 	}
 
-	if r.IOThread != nil {
-		if *r.IOThread {
+	if d.IOThread != nil {
+		if *d.IOThread {
 			values = append(values, "iothread=1")
 		} else {
 			values = append(values, "iothread=0")
 		}
 	}
 
-	if r.SSD != nil {
-		if *r.SSD {
+	if d.SSD != nil {
+		if *d.SSD {
 			values = append(values, "ssd=1")
 		} else {
 			values = append(values, "ssd=0")
 		}
 	}
 
-	if r.Discard != nil && *r.Discard != "" {
-		values = append(values, fmt.Sprintf("discard=%s", *r.Discard))
+	if d.Discard != nil && *d.Discard != "" {
+		values = append(values, fmt.Sprintf("discard=%s", *d.Discard))
 	}
 
-	if r.Cache != nil && *r.Cache != "" {
-		values = append(values, fmt.Sprintf("cache=%s", *r.Cache))
+	if d.Cache != nil && *d.Cache != "" {
+		values = append(values, fmt.Sprintf("cache=%s", *d.Cache))
 	}
 
 	v.Add(key, strings.Join(values, ","))
@@ -1235,8 +1235,8 @@ func (r CustomStorageDevice) EncodeValues(key string, v *url.Values) error {
 }
 
 // EncodeValues converts a CustomStorageDevices array to multiple URL values.
-func (r CustomStorageDevices) EncodeValues(_ string, v *url.Values) error {
-	for s, d := range r {
+func (d CustomStorageDevices) EncodeValues(_ string, v *url.Values) error {
+	for s, d := range d {
 		if d.Enabled {
 			if err := d.EncodeValues(s, v); err != nil {
 				return fmt.Errorf("error encoding storage device %s: %w", s, err)
@@ -1907,7 +1907,7 @@ func (r *CustomStartupOrder) UnmarshalJSON(b []byte) error {
 }
 
 // UnmarshalJSON converts a CustomStorageDevice string to an object.
-func (r *CustomStorageDevice) UnmarshalJSON(b []byte) error {
+func (d *CustomStorageDevice) UnmarshalJSON(b []byte) error {
 	var s string
 
 	if err := json.Unmarshal(b, &s); err != nil {
@@ -1921,24 +1921,24 @@ func (r *CustomStorageDevice) UnmarshalJSON(b []byte) error {
 
 		//nolint:nestif
 		if len(v) == 1 {
-			r.FileVolume = v[0]
+			d.FileVolume = v[0]
 
 			ext := filepath.Ext(v[0])
 			if ext != "" {
 				format := string([]byte(ext)[1:])
-				r.Format = &format
+				d.Format = &format
 			}
 		} else if len(v) == 2 {
 			switch v[0] {
 			case "aio":
-				r.AIO = &v[1]
+				d.AIO = &v[1]
 
 			case "backup":
 				bv := types.CustomBool(v[1] == "1")
-				r.BackupEnabled = &bv
+				d.BackupEnabled = &bv
 
 			case "file":
-				r.FileVolume = v[1]
+				d.FileVolume = v[1]
 
 			case "mbps_rd":
 				iv, err := strconv.Atoi(v[1])
@@ -1946,59 +1946,59 @@ func (r *CustomStorageDevice) UnmarshalJSON(b []byte) error {
 					return fmt.Errorf("failed to convert mbps_rd to int: %w", err)
 				}
 
-				r.MaxReadSpeedMbps = &iv
+				d.MaxReadSpeedMbps = &iv
 			case "mbps_rd_max":
 				iv, err := strconv.Atoi(v[1])
 				if err != nil {
 					return fmt.Errorf("failed to convert mbps_rd_max to int: %w", err)
 				}
 
-				r.BurstableReadSpeedMbps = &iv
+				d.BurstableReadSpeedMbps = &iv
 			case "mbps_wr":
 				iv, err := strconv.Atoi(v[1])
 				if err != nil {
 					return fmt.Errorf("failed to convert mbps_wr to int: %w", err)
 				}
 
-				r.MaxWriteSpeedMbps = &iv
+				d.MaxWriteSpeedMbps = &iv
 			case "mbps_wr_max":
 				iv, err := strconv.Atoi(v[1])
 				if err != nil {
 					return fmt.Errorf("failed to convert mbps_wr_max to int: %w", err)
 				}
 
-				r.BurstableWriteSpeedMbps = &iv
+				d.BurstableWriteSpeedMbps = &iv
 			case "media":
-				r.Media = &v[1]
+				d.Media = &v[1]
 
 			case "size":
-				r.Size = new(types.DiskSize)
-				err := r.Size.UnmarshalJSON([]byte(v[1]))
+				d.Size = new(types.DiskSize)
+				err := d.Size.UnmarshalJSON([]byte(v[1]))
 				if err != nil {
 					return fmt.Errorf("failed to unmarshal disk size: %w", err)
 				}
 
 			case "format":
-				r.Format = &v[1]
+				d.Format = &v[1]
 
 			case "iothread":
 				bv := types.CustomBool(v[1] == "1")
-				r.IOThread = &bv
+				d.IOThread = &bv
 
 			case "ssd":
 				bv := types.CustomBool(v[1] == "1")
-				r.SSD = &bv
+				d.SSD = &bv
 
 			case "discard":
-				r.Discard = &v[1]
+				d.Discard = &v[1]
 
 			case "cache":
-				r.Cache = &v[1]
+				d.Cache = &v[1]
 			}
 		}
 	}
 
-	r.Enabled = true
+	d.Enabled = true
 
 	return nil
 }

--- a/proxmox/nodes/vms/vms_types_test.go
+++ b/proxmox/nodes/vms/vms_types_test.go
@@ -9,6 +9,7 @@ package vms
 import (
 	"testing"
 
+	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 
 	"github.com/bpg/terraform-provider-proxmox/proxmox/types"
@@ -61,6 +62,60 @@ func TestCustomStorageDevice_UnmarshalJSON(t *testing.T) {
 				t.Errorf("UnmarshalJSON() error = %v, wantErr %v", err, tt.wantErr)
 			}
 			require.Equal(t, tt.want, r)
+		})
+	}
+}
+
+func TestCustomStorageDevice_IsCloudInitDrive(t *testing.T) {
+	t.Parallel()
+
+	tests := []struct {
+		name   string
+		device CustomStorageDevice
+		want   bool
+	}{
+		{
+			name: "simple volume",
+			device: CustomStorageDevice{
+				FileVolume: "local-lvm:vm-131-disk-0",
+			},
+			want: false,
+		}, {
+			name: "on directory storage",
+			device: CustomStorageDevice{
+				Media:      types.StrPtr("cdrom"),
+				FileVolume: "local:131/vm-131-cloudinit.qcow2",
+			},
+			want: true,
+		}, {
+			name: "on block storage",
+			device: CustomStorageDevice{
+				Media:      types.StrPtr("cdrom"),
+				FileVolume: "local-lvm:vm-131-cloudinit",
+			},
+			want: true,
+		}, {
+			name: "wrong VM ID",
+			device: CustomStorageDevice{
+				Media:      types.StrPtr("cdrom"),
+				FileVolume: "local-lvm:vm-123-cloudinit",
+			},
+			want: false,
+		}, {
+			name: "not a cdrom",
+			device: CustomStorageDevice{
+				FileVolume: "local-lvm:vm-123-cloudinit",
+			},
+			want: false,
+		},
+	}
+
+	for _, tt := range tests {
+		tt := tt
+		t.Run(tt.name, func(t *testing.T) {
+			t.Parallel()
+			got := tt.device.IsCloudInitDrive(131)
+			assert.Equal(t, tt.want, got)
 		})
 	}
 }

--- a/proxmoxtf/resource/vm.go
+++ b/proxmoxtf/resource/vm.go
@@ -1662,10 +1662,7 @@ func findExistingCloudInitDrive(vmConfig *vms.GetResponseData, vmID int, default
 		vmConfig.IDEDevice0, vmConfig.IDEDevice1, vmConfig.IDEDevice2, vmConfig.IDEDevice3,
 	}
 	for i, device := range devices {
-		if device != nil && device.Enabled && device.Media != nil && *device.Media == "cdrom" && strings.Contains(
-			device.FileVolume,
-			fmt.Sprintf("vm-%d-cloudinit", vmID),
-		) {
+		if device != nil && device.Enabled && device.IsCloudInitDrive(vmID) {
 			return fmt.Sprintf("ide%d", i)
 		}
 	}
@@ -4110,7 +4107,7 @@ func vmReadCustom(
 			continue
 		}
 
-		if strings.HasSuffix(dd.FileVolume, fmt.Sprintf("vm-%d-cloudinit", vmID)) {
+		if dd.IsCloudInitDrive(vmID) {
 			continue
 		}
 

